File: cpp/306.cpp
#include <string>
#include <iostream>
using namespace std;

int len(long long n){
  int l = 1;
  long long t = 10;
  while(1){
    if(n < t)
      return l;
    l += 1;
    t *= 10;
  }
}

class Solution {
public:
  bool isAdditiveNumber(string num) {
    int n = num.size();
    if(n < 3)
      return false;
    long long n1 = 0;
    for(int i1 = 0; i1 < n-2; i1++){
      n1 = n1 * 10 + num[i1] - 48;
      if(i1 > 0 && num[0] == '0')
        return false;
      long long n2 = 0;
      for(int i2 = i1+1; i2 < n-1; i2++){
        n2 = n2 * 10 + num[i2] - 48;
        if(i2 - i1 > 1 && num[i1+1] == '0')
          continue;
        int i3 = i2 + 1;
        long long pre = n2, n3 = n1 + n2;
        int l = len(n3);
        while(i3 + l <= n){
          long long t3 = 0;
          for(int j=i3; j<i3+l; j++)
            t3 = 10 * t3 + num[j] - 48;
          if(t3 != n3)
            break;
          n3 = pre + n3;
          pre = t3;
          i3 += l;
          l = len(n3);
        }
        if(i3 == n)
          return true;
      }
    }
    return false;
  }
};

int main(){
  cout << Solution().isAdditiveNumber("121474836472147483648");
}

File: python/306.py
class Solution(object):
    @staticmethod
    def len(n):
        l = 1
        t = 10
        while True:
            if n < t:
                return l
            t *= 10
            l += 1

    def isAdditiveNumber(self, num):
        """
        :type num: str
        :rtype: bool
        """
        n = len(num)
        if n < 3:
            return False
        n1 = 0
        for i1 in range(n-2):
            n1 = 10 * n1 + int(num[i1])
            if i1 > 0 and num[0] == '0':
                return False
            n2 = 0
            for i2 in range(i1+1, n-1):
                n2 = 10 * n2 + int(num[i2])
                if i2 - i1 > 1 and num[i1+1] == '0':
                    continue
                i3 = i2 + 1
                pre = n2
                n3 = n1 + n2
                l = self.len(n3)
                while i3 + l <= n:
                    if num[i3:i3+l] != str(n3):
                        break
                    pre, n3 = n3, pre + n3
                    i3 += l
                    l = self.len(n3)
                if i3 == n:
                    print(n1, n2)
                    return True
        return False

s = Solution()
print(s.isAdditiveNumber("121474836472147483648"))
